### PR TITLE
Test fixes (node 12 and workflows reorganization)

### PR DESCRIPTION
**Pull request by tekwiz**
**What is the purpose of this pull request?**

- [x] Documentation update
- [x] Bug fix
- [x] New feature
- [x] Other, please explain: Reworks CI test workflows

**What changes did you make? (provide an overview)**

1. Adds `c8 ignore` directives to node-version-specific code (2d7158d610d1f896250e5b6a6c39492ac8e07595) and fixes the directive at the end of the Headers constructor (f5f58fb3890494fa2f531ba8d9191eeb7ad8880b)
2. Fixes the test-artifact script so it will run on Windows (b914a9b26fdb813a4a0f1d0c973d0f5ca69897c6)
3. Changes the `prepublishOnly` script to `test:cjs-build` and adds it to the main ci workflow so it is tested on all environments (6d5769007a29846a59a5bb50d9e20fdcdc13ce2c)
4. Fixes a problem with using port 5000 in tests on shared CI environments, i.e. GitHub Actions (d4490ec0059594d00b0c7ff4b6dc520f9b27108e).  You can see where I hit this [in GitHub action run 1470398827](https://github.com/tekwiz/node-fetch/runs/1470398827).
5. Fixes a bug with the Node 12.0-12.9 version comparison for the aborted connection fix (#1038, 122614be9086697904864ee8e2f98f91e7b61325)
6. Fixes test running on Node 12 and reorganizes the GitHub workflows to test necessary environments. (878498da5002d4dd4d0d37755700150506efd5e2 and daed7f43b2d0ded6db4b82e6355481b17ca3bd60)

**Which issue (if any) does this pull request address?**

* #1038 - Question about comparing Node version bug (fixed in 122614be9086697904864ee8e2f98f91e7b61325)

**Is there anything you'd like reviewers to know?**

This all started when I was trying to get the tests to run on Node 12.14 because that is the 12.x version that runs on AWS Lambda.  While working on this, I wanted to ensure test coverage was returned to 100% so I knew I had no missing pieces.  In that process, I encountered several issues, some of which are fixed separately (#1018, #1019, #1020, and #1022).  The main issue that this PR addresses is around the conversion to ECMAScript modules.  You will note the number of Node.js versions in the workflow strategy has increased.  Here is why:

* 10.17.0: lowest allowed version (package.json)
* 12.3.0: lowest allowed 12.x version (package.json)
* 12.16.3: [esm](https://www.npmjs.com/package/esm) package doesn't work properly with node 12.16, `--experimental-modules` is required
* 12.x (latest): `--experimental-modules` is on by default starting in 12.17.0 (see [Node.js Blog: Node v12.17.0 (LTS), "ECMAScript Modules - `--experimental-modules` flag removal", 2020 May 26](
  https://nodejs.org/en/blog/release/v12.17.0/#ecmascript-modules-experimental-modules-flag-removal))
* 14.x: Active LTS

Also, it is worth noting that combining #1017, #1018, #1019, #1020, #1022, and this pull request results in 100% coverage:

- [Coveralls report: tekwiz:multiple/20201213](https://coveralls.io/github/tekwiz/node-fetch?branch=multiple/20201213)
- [GitHub action result: typescript](https://github.com/tekwiz/node-fetch/actions/runs/419088164)
- [GitHub action result: lint](https://github.com/tekwiz/node-fetch/actions/runs/419088165)
- [GitHub action result: test](https://github.com/tekwiz/node-fetch/actions/runs/419088166)

Last, but certainly not least -- maintainers should read daed7f43b2d0ded6db4b82e6355481b17ca3bd60 for how running tests changes with this.

### CODE REVIEW

**Review by xxczaki — APPROVED** [review]

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -1,10 +1,11 @@
 name: CI
 
 on:
+  push:
+    branches: [master]
   pull_request:
     paths:
       - "**.js"
-      - "**eslint**"
```
Comment (tinovyatkin) [inline .github/workflows/lint.yml:7]: Why do you removing this? It supposes to run lint what ESlint configuration was changed

File: test/commonjs/test-artifact.js [hunk 2]
```diff
@@ -2,10 +2,10 @@
 /**
  * Rebuild first
  */
-const {execFileSync} = require('child_process');
+const {execSync} = require('child_process');
 
 console.log('Building CommonJS version...');
-execFileSync('npm', ['run', 'build'], {stdio: 'inherit'});
+execSync('npm run build', {stdio: 'inherit'});
```
Comment (tinovyatkin) [inline test/commonjs/test-artifact.js:8]: Can you elaborate on why this change is required?

File: src/index.js [hunk 3]
```diff
@@ -187,7 +187,8 @@ export default async function fetch(url, options_) {
 
 			let body = pump(response_, new PassThrough(), reject);
 			// see https://github.com/nodejs/node/pull/29376
-			if (process.version < 'v12.10') {
+			/* c8 ignore next 3 */
+			if (/^v12\.\d\./.test(process.version) || process.version < 'v12') { // i.e. semver <12.10
```
Comment (tinovyatkin) [inline src/index.js:191]: Sorry, I didn't get this change. It was comparing with 12.10 earlier and now it's just 12?

File: package.json [hunk 4]
```diff
@@ -24,11 +24,16 @@
...
+    "test": "c8 --reporter=html --reporter=lcov --reporter=text --check-coverage npm run test:nocoverage",
+    "test:experimental:nocoverage": "mocha --experimental-modules --colors",
+    "test:experimental": "c8 --reporter=html --reporter=lcov --reporter=text --check-coverage npm run test:experimental:nocoverage",
+    "test:esm:nocoverage": "mocha -r esm --colors",
+    "test:esm": "c8 --reporter=html --reporter=lcov --reporter=text --check-coverage npm run test:esm:nocoverage",
+    "test:cjs-build": "node ./test/commonjs/test-artifact.js",
     "coverage": "c8 report --reporter=text-lcov | coveralls",
     "test-types": "tsd",
-    "lint": "xo",
-    "prepublishOnly": "node ./test/commonjs/test-artifact.js"
```
Comment (tinovyatkin) [inline package.json:31]: Why we don't want to test this anymore?

**Review by tinovyatkin — CHANGES_REQUESTED** [review]
Why hardcoded minimal Node version is better than pulling that from `package.json`?

File: .github/workflows/lint.yml (hunk 1 above)
Comment (tekwiz) [reply .github/workflows/lint.yml:7]: @tinovyatkin Because there are no more eslint configuration files.  Everything appears to have been moved to xo and package.json

File: test/commonjs/test-artifact.js (hunk 2 above)
Comment (tekwiz) [reply test/commonjs/test-artifact.js:8]: @tinovyatkin `npm` is a batch file on Windows, which can't be executed by `execFile`.

File: src/index.js (hunk 3 above)
Comment (tekwiz) [reply src/index.js:191]: @tinovyatkin See #1038 for the problem. The solution, here, is the regex, which matches for a **single-digit** minor version of 12, i.e. 12.0.x - 12.9.x. I know it's obtuse, but it seemed more reasonable than adding the  whole semver package just for this one condition.

File: package.json (hunk 4 above)
Comment (tekwiz) [reply package.json:31]: @tinovyatkin We do want it. It's just been renamed to `test:cjs-build` since it's not just a pre-publish step.

**Comment by tekwiz** [comment]
@tinovyatkin I've made the change to the CI.yml to pull the minimum node-version from `package.json`

File: package.json (hunk 4 above)
Comment (tinovyatkin) [reply package.json:31]: But should we still call this from prepublishOnly? Why we are completely removing prepublish check?

File: .github/workflows/lint.yml (hunk 1 above)
Comment (tinovyatkin) [reply .github/workflows/lint.yml:7]: Ok, but how it hurts to keep this line? If some contributor will (for some reason) add `.eslintignore` or `.eslintrc` we would like this workflow to run, don't we?

File: src/index.js (hunk 3 above)
Comment (tinovyatkin) [reply src/index.js:191]: Hmm... Can we construct a RegEx then to cover these conditions without the need for a second check? I mean, it looks redundant to first check the line with RegEx and then with string comparison.... Something like `/^v1([01]|2\.\d)\.d/` should work in a single pass, right (our minimum supported version is 10.x already)?

Comment (tekwiz) [reply src/index.js:191]: @tinovyatkin Sure, I can do that. I thought was I had done here was simpler, but it seems I was wrong 🙂. I'll make this change. M

File: .github/workflows/lint.yml (hunk 1 above)
Comment (tekwiz) [reply .github/workflows/lint.yml]: @tinovyatkin Good point. I've made this change

File: package.json (hunk 4 above)
Comment (tekwiz) [reply package.json:31]: @tinovyatkin Ah, you are quite right. I've fixed this.

File: src/index.js (hunk 3 above)
Comment (tekwiz) [reply src/index.js]: This change is made.

**Review by tinovyatkin — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -725,6 +725,20 @@ For older versions please use the type definitions from [DefinitelyTyped](https:
 npm install --save-dev @types/node-fetch
 ```
 
+## ECMAScript modules -- development and tests
```
Comment (Richienb) [inline README.md:728]: This should go in a `contributing.md`.

**Review by Richienb — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (tekwiz) [reply README.md]: @Richienb I've started a CONTRIBUTING.md file and moved this section there

**Comment by tekwiz** [comment]
@Richienb Are there any other changes you would like to see on this?

**Comment by tekwiz** [comment]
Superseded; closing.
